### DISCUSSION TITLE
Utility types?

### DISCUSSION THREAD

**Discussion by djMax** [openapi-ts/openapi-typescript#1259] [Ideas]
I find this helper, well... helpful - would a PR for them be welcome?

```
export type JsonRequestBody<T, M extends keyof T> = T[M] extends {
  requestBody: { content: { 'application/json': infer U } };
}
  ? U
  : never;
``

**Comment by drwpow** [comment]
Ha so I’ve been playing around with some other derived stuff and have also found myself needing similar helpers as well. 🤔 I wonder now that this is a monorepo if we just make a third package that’s just any helper utilities that are remotely useful. That way it’s “free” to install because they’re only TS types (someone may not be using openapi-typescript, for example, but want some helpers).

**Comment by djMax** [comment]
Yes, I agree that these are slightly different because they have no runtime impact. I'd argue they would be best in openapi-typescript because even if you don't want to use the generation, the penalty is minimal, and it would be a devDependency anyways, but I can certainly see the other perspective. The other thing I really want is "real" enums, and that DOES have a runtime impact so it's more complex. (This isn't supported right now right?)

**Reply by drwpow** [reply]
No, generating enums isn’t supported yet, but there is an open ticket for it and I’d be open to that landing

**Reply by drwpow** [reply]
And also, after sitting on it for a few days, I think I’d prefer its own package for this, because openapi-typescript doesn’t really _need_ these types per-se. And the versioning may exist completely independent of `openapi-typescript` (e.g. what if you wanted type from version X but CLI version Y?).

If anything, `openapi-fetch` would depend on this new type package, but that’d be OK.

**Comment by djMax** [comment]
I did the enum thing [here](/openapi-typescript-infra/openapi-typescript-enum). What might be an even more flexible change is just to allow transform/postTransform to be specified on the CLI.

**Reply by drwpow** [reply]
Oh that’s right. I recall you outlining this in another issue as well. This is something I’ll probably take a look at this month.

**Comment by drwpow** [comment]
Semi-related: just remembered [this issue](https://github.com/drwpow/openapi-typescript/issues/922) that outlines a type helper someone needed.

**Comment by drwpow** [comment]
Just merged #1300 which adds the `openapi-typescript-helpers` package, a types-only package for these kinds of things. Right now it’s just one big dump of types, but over time I’d have no problem expanding it to include anything generally-useful (also, feel free to submit any PRs to add any types!)

**Comment by drwpow** [comment]
With `openapi-typescript-helpers` now available, any/all PRs to that are welcome! If something is OpenAPI-adjacent and would be helpful to expose there, there’s no harm in adding it. Going to close this discussion but feel free to leave any additional thoughts here or in PRs.
